components/site/src/link_checking.rs

```diff
@@ -3,84 +3,68 @@ use rayon::prelude::*;
     let page_links = library
         .pages()
         .values()
         .map(|p| {
             let path = &p.file.path;
-            p.internal_links_with_anchors.iter().map(move |l| (path.clone(), l))
+            p.internal_links.iter().map(move |l| (path.clone(), l))
         })
         .flatten();
     let section_links = library
         .sections()
         .values()
         .map(|p| {
             let path = &p.file.path;
-            p.internal_links_with_anchors.iter().map(move |l| (path.clone(), l))
+            p.internal_links.iter().map(move |l| (path.clone(), l))
         })
         .flatten();
-    let all_links = page_links.chain(section_links).collect::<Vec<_>>();
-
-    if site.config.is_in_check_mode() {
-        println!("Checking {} internal link(s) with an anchor.", all_links.len());
-    }
-
-    if all_links.is_empty() {
-        return Ok(());
-    }
-
-    let mut full_path = site.base_path.clone();
-    full_path.push("content");
-
-    let errors: Vec<_> = all_links
-        .iter()
-        .filter_map(|(page_path, (md_path, anchor))| {
-            // There are a few `expect` here since the presence of the .md file will
-            // already have been checked in the markdown rendering
-            let mut p = full_path.clone();
-            for part in md_path.split('/') {
-                p.push(part);
-            }
-            if md_path.contains("_index.md") {
-                let section = library
-                    .get_section(&p)
-                    .expect("Couldn't find section in check_internal_links_with_anchors");
-                if section.has_anchor(&anchor) {
-                    None
-                } else {
-                    Some((page_path, md_path, anchor))
-                }
-            } else {
-                let page = library
-                    .get_page(&p)
-                    .expect("Couldn't find section in check_internal_links_with_anchors");
-                if page.has_anchor(&anchor) {
-                    None
-                } else {
-                    Some((page_path, md_path, anchor))
-                }
-            }
+    let all_links = page_links.chain(section_links);
+
+    // Only keep links with anchor fragments, and count them too.
+    // Bare files have already been checked elsewhere, thus they are not interesting here.
+    let mut anchors_total = 0usize;
+    let links_with_anchors = all_links
+        .filter_map(|(page_path, link)| match link {
+            (md_path, Some(anchor)) => Some((page_path, md_path, anchor)),
+            _ => None,
         })
-        .collect();
-
-    if site.config.is_in_check_mode() {
-        println!(
-            "> Checked {} internal link(s) with an anchor: {} error(s) found.",
-            all_links.len(),
-            errors.len()
-        );
-    }
-
-    if errors.is_empty() {
-        return Ok(());
-    }
+        .inspect(|_| anchors_total = anchors_total.saturating_add(1));
```

Comment:
TIL about `inspect`. Do we expect to have more than u64::MAX_SIZE anchors?
Is it not as simple as doing `links_with_anchors.len()`?

Reply:
> Do we expect to have more than u64::MAX_SIZE anchors?

Quite unlikely. The `saturating_add` only makes the arithmetic explicit. If it feels un-ergonomic in this codebase, it can be changed to a `+= 1`, whose behavior depends on build-time configuration (however neither panic-ing nor overflowing is really useful here IMHO).

> Is it not as simple as doing links_with_anchors.len()?

Not really, mostly because `links_with_anchors` does not exist as a collection. If you look at the rest of the changes to this function, you'll notice that it used to iterate multiple times over the input and collect into two intermediate vectors.
This instead builds a single iterator and performs a single collection at the end, however it has to count elements on the fly.